File: cpp11/Ice/asyncInvocation/CalculatorI.cpp
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <CalculatorI.h>
#include <math.h>

using namespace std;

int
CalculatorI::add(int x, int y, const Ice::Current&)
{
    return x + y;
}

int
CalculatorI::subtract(int x, int subtrahend, const Ice::Current&)
{
    return x - subtrahend;
}

int
CalculatorI::divide(int numerator, int denominator, int& remainder, const Ice::Current&)
{
    if(denominator == 0)
    {
        throw Demo::DivideByZeroException();
    }
    remainder = numerator % denominator;
    return numerator / denominator;
}

int
CalculatorI::square(int x, const Ice::Current&)
{
    return x * x;
}

double
CalculatorI::squareRoot(int x, const Ice::Current&)
{
    if(x < 0)
    {
        throw Demo::NegativeRootException();
    }
    return sqrt(x);
}

void
CalculatorI::shutdown(const Ice::Current& current)
{
    current.adapter->getCommunicator()->shutdown();
}

File: swift/Ice/asyncInvocation/Sources/Server/CalculatorI.swift
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Ice

class CalculatorI: Calculator {
    func add(x: Int32, y: Int32, current _: Ice.Current) throws -> Int32 {
        return x + y
    }

    func subtract(x: Int32, subtrahend: Int32, current _: Ice.Current) throws -> Int32 {
        return x - subtrahend
    }

    func divide(numerator: Int32,
                denominator: Int32,
                current _: Ice.Current) throws -> (returnValue: Int32, remainder: Int32) {
        guard denominator != 0 else {
            throw DivideByZeroException()
        }
        return (numerator / denominator, numerator % denominator)
    }

    func square(x: Int32, current _: Ice.Current) throws -> Int32 {
        return x * x
    }

    func squareRoot(x: Int32, current _: Ice.Current) throws -> Double {
        guard x >= 0 else {
            throw NegativeRootException()
        }
        return Double(x).squareRoot()
    }

    func shutdown(current: Ice.Current) {
        current.adapter!.getCommunicator().shutdown()
    }
}
